server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,298 @@
+          cacheKeyStr,
+          newResultSetId
+      );
+
+      responseContext.remove(QueryResource.EXISTING_RESULT_ID);
+      if (newResultSetId != null && newResultSetId.equals(existingResultSetId)) {
+        log.info("Return cached result set as there is no change in identifiers for query %s ", query.getId());
+        return deserializeResults(cachedResultSet, strategy, existingResultSetId);
+      } else {
+        return Sequences.wrap(Sequences.map(
+            resultFromClient,
+            new Function<T, T>()
+            {
+              @Override
+              public T apply(T input)
+              {
+                cacheResultEntry(resultLevelCachePopulator, input);
+                return input;
+              }
+            }
+        ), new SequenceWrapper()
+        {
+          @Override
+          public void after(boolean isDone, Throwable thrown) throws Exception
+          {
+            if (resultLevelCachePopulator != null) {
+              // The resultset identifier and its length is cached along with the resultset
+              resultLevelCachePopulator.populateResults(newResultSetId);
+              log.info("Cache population complete for query %s", query.getId());
+            }
+          }
+        });
+      }
+    } else {
+      return baseRunner.run(
+          queryPlus,
+          responseContext
+      );
+    }
+  }
+
+  private T cacheResultEntry(
+      ResultLevelCachePopulator resultLevelCachePopulator,
+      T resultEntry
+  )
+  {
+    final Function<T, Object> cacheFn = strategy.prepareForResultLevelCache();
+    if (resultLevelCachePopulator != null) {
+      resultLevelCachePopulator.cacheFutures
+          .add(backgroundExecutorService.submit(() -> cacheFn.apply(resultEntry)));
+    }
+    return resultEntry;
+  }
+
+  private byte[] fetchResultsFromResultLevelCache(
+      final String queryCacheKey
+  )
+  {
+    if (useResultCache && queryCacheKey != null) {
+      return cache.get(ResultLevelCacheUtil.computeResultLevelCacheKey(queryCacheKey));
+    }
+    return null;
+  }
+
+  private String extractEtagFromResults(
+      final byte[] cachedResult
+  )
+  {
+    if (cachedResult == null) {
+      return null;
+    }
+    log.info("Fetching result level cache identifier for query: %s", query.getId());
+    int etagLength = ByteBuffer.wrap(cachedResult, 0, Integer.BYTES).getInt();
+    return StringUtils.fromUtf8(Arrays.copyOfRange(cachedResult, Integer.BYTES, etagLength + Integer.BYTES));
+  }
+
+  private Sequence<T> deserializeResults(
+      final byte[] cachedResult, CacheStrategy strategy, String resultSetId
+  )
+  {
+    if (cachedResult == null) {
+      return null;
+    }
+    final Function<Object, T> pullFromCacheFunction = strategy.pullFromResultLevelCache();
+    final TypeReference<Object> cacheObjectClazz = strategy.getCacheObjectClazz();
+    //Skip the resultsetID and its length bytes
+    byte[] prunedCacheData = Arrays.copyOfRange(
```

Comment:
we must avoid this, this just created another copy of whole result set in memory again